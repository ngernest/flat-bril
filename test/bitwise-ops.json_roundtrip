{
  "functions": [
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "a",
            "two"
          ],
          "dest": "tmp",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "tmp",
            "two"
          ],
          "dest": "tmp2",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a",
            "tmp2"
          ],
          "dest": "tmp3",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "one",
            "tmp3"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "ans"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod2",
      "type": "bool"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        },
        {
          "name": "c",
          "type": "bool"
        }
      ],
      "instrs": [
        {
          "dest": "i",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "n",
          "op": "const",
          "type": "int",
          "value": 63
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "ans",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "to_add",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "label": "loop"
        },
        {
          "args": [
            "i",
            "n"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "funcs": [],
          "labels": [
            "here",
            "end"
          ],
          "op": "br"
        },
        {
          "label": "here"
        },
        {
          "args": [
            "a"
          ],
          "dest": "mod2a",
          "funcs": [
            "mod2"
          ],
          "labels": [],
          "op": "call",
          "type": "bool"
        },
        {
          "args": [
            "b"
          ],
          "dest": "mod2b",
          "funcs": [
            "mod2"
          ],
          "labels": [],
          "op": "call",
          "type": "bool"
        },
        {
          "args": [
            "mod2a",
            "mod2b"
          ],
          "dest": "cond_add",
          "funcs": [],
          "labels": [],
          "op": "and",
          "type": "bool"
        },
        {
          "args": [
            "c"
          ],
          "funcs": [],
          "labels": [
            "doOr",
            "stay"
          ],
          "op": "br"
        },
        {
          "label": "doOr"
        },
        {
          "args": [
            "mod2a",
            "mod2b"
          ],
          "dest": "cond_add",
          "funcs": [],
          "labels": [],
          "op": "or",
          "type": "bool"
        },
        {
          "label": "stay"
        },
        {
          "args": [
            "cond_add"
          ],
          "funcs": [],
          "labels": [
            "add",
            "end_loop"
          ],
          "op": "br"
        },
        {
          "label": "add"
        },
        {
          "args": [
            "ans",
            "to_add"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "label": "end_loop"
        },
        {
          "args": [
            "a",
            "two"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "b",
            "two"
          ],
          "dest": "b",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "to_add",
            "two"
          ],
          "dest": "to_add",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "i",
            "one"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loop"
          ],
          "op": "jmp"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "ans"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "loop_subroutine",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "oper",
          "op": "const",
          "type": "bool",
          "value": true
        },
        {
          "args": [
            "a",
            "b",
            "oper"
          ],
          "dest": "v1",
          "funcs": [
            "loop_subroutine"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v1"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "OR",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "oper",
          "op": "const",
          "type": "bool",
          "value": false
        },
        {
          "args": [
            "a",
            "b",
            "oper"
          ],
          "dest": "v1",
          "funcs": [
            "loop_subroutine"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v1"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "AND",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "and_val",
          "funcs": [
            "AND"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "or_val",
          "funcs": [
            "OR"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "or_val",
            "and_val"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "ans"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "XOR",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        },
        {
          "name": "c",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "c",
            "one"
          ],
          "dest": "sel",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "zero"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "sel",
            "zero"
          ],
          "dest": "less",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "sel",
            "zero"
          ],
          "dest": "equal",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "sel",
            "zero"
          ],
          "dest": "greater",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "less"
          ],
          "funcs": [],
          "labels": [
            "and_op",
            "useless_lbl"
          ],
          "op": "br"
        },
        {
          "label": "useless_lbl"
        },
        {
          "args": [
            "equal"
          ],
          "funcs": [],
          "labels": [
            "or_op",
            "xor_op"
          ],
          "op": "br"
        },
        {
          "label": "and_op"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "ans",
          "funcs": [
            "AND"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "or_op"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "ans",
          "funcs": [
            "OR"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "end"
          ],
          "op": "jmp"
        },
        {
          "label": "xor_op"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "ans",
          "funcs": [
            "XOR"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "label": "end"
        },
        {
          "args": [
            "ans"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
