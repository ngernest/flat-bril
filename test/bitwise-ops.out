@mod2
	op: const		dest: two		type: int		value: 2	
	op: div  		dest: tmp		type: int		args: ["a", "two"]	
	op: mul  		dest: tmp2		type: int		args: ["tmp", "two"]	
	op: sub  		dest: tmp3		type: int		args: ["a", "tmp2"]	
	op: const		dest: one		type: int		value: 1	
	op: eq   		dest: ans		type: bool		args: ["one", "tmp3"]	
	op: ret  		args: ["ans"]	
@loop_subroutine
	op: const		dest: i 		type: int		value: 0	
	op: const		dest: n 		type: int		value: 63	
	op: const		dest: one		type: int		value: 1	
	op: const		dest: two		type: int		value: 2	
	op: const		dest: ans		type: int		value: 0	
	op: const		dest: to_add		type: int		value: 1	
.loop
	op: le   		dest: cond		type: bool		args: ["i", "n"]	
	op: br   		args: ["cond"]		labels: ["here", "end"]
.here
	op: call 		dest: mod2a		type: bool		args: ["a"]		funcs: ["mod2"]
	op: call 		dest: mod2b		type: bool		args: ["b"]		funcs: ["mod2"]
	op: and  		dest: cond_add		type: bool		args: ["mod2a", "mod2b"]	
	op: br   		args: ["c"]		labels: ["doOr", "stay"]
.doOr
	op: or   		dest: cond_add		type: bool		args: ["mod2a", "mod2b"]	
.stay
	op: br   		args: ["cond_add"]		labels: ["add", "end_loop"]
.add
	op: add  		dest: ans		type: int		args: ["ans", "to_add"]	
.end_loop
	op: div  		dest: a 		type: int		args: ["a", "two"]	
	op: div  		dest: b 		type: int		args: ["b", "two"]	
	op: mul  		dest: to_add		type: int		args: ["to_add", "two"]	
	op: add  		dest: i 		type: int		args: ["i", "one"]	
	op: jmp  		labels: ["loop"]
.end
	op: ret  		args: ["ans"]	
@OR
	op: const		dest: oper		type: bool		value: true	
	op: call 		dest: v1		type: int		args: ["a", "b", "oper"]		funcs: ["loop_subroutine"]
	op: ret  		args: ["v1"]	
@AND
	op: const		dest: oper		type: bool		value: false	
	op: call 		dest: v1		type: int		args: ["a", "b", "oper"]		funcs: ["loop_subroutine"]
	op: ret  		args: ["v1"]	
@XOR
	op: call 		dest: and_val		type: int		args: ["a", "b"]		funcs: ["AND"]
	op: call 		dest: or_val		type: int		args: ["a", "b"]		funcs: ["OR"]
	op: sub  		dest: ans		type: int		args: ["or_val", "and_val"]	
	op: ret  		args: ["ans"]	
@main
	op: const		dest: one		type: int		value: 1	
	op: const		dest: zero		type: int		value: 0	
	op: sub  		dest: sel		type: int		args: ["c", "one"]	
	op: id   		dest: ans		type: int		args: ["zero"]	
	op: lt   		dest: less		type: bool		args: ["sel", "zero"]	
	op: eq   		dest: equal		type: bool		args: ["sel", "zero"]	
	op: gt   		dest: greater		type: bool		args: ["sel", "zero"]	
	op: br   		args: ["less"]		labels: ["and_op", "useless_lbl"]
.useless_lbl
	op: br   		args: ["equal"]		labels: ["or_op", "xor_op"]
.and_op
	op: call 		dest: ans		type: int		args: ["a", "b"]		funcs: ["AND"]
	op: jmp  		labels: ["end"]
.or_op
	op: call 		dest: ans		type: int		args: ["a", "b"]		funcs: ["OR"]
	op: jmp  		labels: ["end"]
.xor_op
	op: call 		dest: ans		type: int		args: ["a", "b"]		funcs: ["XOR"]
.end
	op: print		args: ["ans"]	
